### DISCUSSION TITLE
Automerge without actor IDs

### DISCUSSION THREAD

**Discussion by alexjg** [automerge/automerge#546] [Ideas]
I would like to get rid of actor IDs, I think they are currently a bit of a footgun. Our current documentation doesn't talk about actor IDs much and in practice users generate a new actor ID for each run of a program. The problem with this is that for long lived documents (or documents with many editors) this can lead to an unexpected explosion in the size of the encoded document because we must encode every single actor in the compressed document. This problem is likely to become much worse once we support branches, because users will also need to create a new actor for each branch they edit.

One way to deal with this is to build much more tooling for managing actor IDs and to explain and emphasize management of actor IDs in the documentation. Maybe that is the route we will end up going down, but first I wanted to think about what automerge might look like without actor IDs at all.

Actor IDs seem to me to be important because of these properties of operation IDs built on actor IDs:

1. They encode quite succintly in the columnar document format because for runs of consecutive inserts (i.e. text) the counter component delta encodes to a single bit per op, which then RLE encodes to a single integer for the run and the actor ID (which is constant for the run) RLE encodes to a single integer for the whole run. 
2. They are straightforward to build an order over, the operation ID is a lamport timestamp

Now, imagine if we were to generate an actor ID for every new change to the document. If every change has a unique actor ID then instead of generating a new actor ID, we could just use the change hash as the actor ID and the above properties would be preserved. One difference with the current system is that two nodes could produce the same change (if they used the same timestamp). This seems like it shouldn't be a problem but maybe I have missed something?

## Encoding

If we naively just replaced actor IDs with change hashes everywhere in the current system we wouldn't be any better off. The advantage of using change hashes is that we already have a representation of the change graph so we don't need to encode the hashes, this allows us to remove the exploding actor IDs footgun. We do still need to come up with a way to encode the change hash component of an op ID succinctly. Here are some ideas I have had:

* When encoding ops in document blocks the "actor" column currently is an integer index into the actor IDs encoded at the start of the document. The document block also includes a list of changes (which is the non-operation based metadata in a change) so we could replace the actor column with a "change" column, which is index into this list of changes
* When encoding ops in change blocks we have a few options:
    * Encode the change hashes referred to by any ops in the change at the start of the block and refer to those changes in the ops (we already encode the direct dependencies in the change, we would now have to encode any transitive dependencies referred to by an op in the change). This is what we currently do with actors. This might not be as effective as current compression though because if there are many ancestor operations split over many ancestor changes which were all notionally by the same actor then you would end up with more change hashes than actor IDs
    * Encode the change hashes as indexes into the topological sort of the transitive dependencies of the change. This would be very succinct but would be computationally expensive to reconstruct (you have to realize the topological sort of the dependencies of every change in the change graph). 

## Ordering

In the internals of automerge we rely on being able to compare two op IDs easily. The change I am suggesting here doesn't make the comparison of two op IDs much different, what does change is our use of vector clocks. We use vector clocks whenever we want to filter ops down to a particular point in the change graph - such as when querying for the values of a key at a given set of heads. I think this is tractable though, instead of using a vector clock to represent a point in time we would just use something like the set of transitive dependencies of the given heads and then we know that an operation ID is visible if its change hash is in the set (we would use some kind of index into the change hash set rather than an actual hash, which is what we already do with actor IDs).

**Comment by ept** [comment]
Great analysis! I have also long wanted to get rid of actor IDs, and I came to similar conclusions as you did. The obstacles are probably surmountable, but they introduce more implementation complexity, so I think it was right to ship the first version of the binary encoding with both hash chaining and actor IDs, even though they are essentially two different ways of expressing the same thing. Now that we have that working well, it makes sense to start planning a v2 format that removes actor IDs.

## Encoding individual changes

For encoding ops into individual changes, explicitly listing the hash of every single opID that's mentioned in the change could make changes much bigger. For example, take a paragraph of text that was typed one character at a time, so each character's ID is in a different change. If that whole paragraph is subsequently deleted, we generate a change with an individual deletion op for each character in the paragraph. With actor IDs this compresses well, but with hashes, the size of this deletion change would be >32 bytes times the number of deleted chars. A few possible responses to this concern:

1. Declare that it's not a problem because in text editing, many-char deletion changes occur rarely compared to single-char changes, and so it doesn't matter if they're a bit bigger. The deletion should still be smaller than the log of insertion ops that created the deleted paragraph in the first place. (However, we'd need to think carefully about how this interacts with our ideas for compressing blocks of changes.)
2. Support a single op that deletes a whole range of chars at once, as per automerge/automerge-classic#401. That's probably a good idea anyway, but orthogonal to this proposal.
3. Encoding the change hashes as indexes into the topological sort of the transitive dependencies of the change, as you propose: a simplification of this would be to support indexing only into a linear sequence of changes that each have exactly one dependency (which is the common case). When we come across a change with multiple dependencies, we can't index beyond it, but have to give the explicit hash of the dependency.

## Ordering

For ordering, it's important that all peers agree on the relative ordering of opIDs (for example, to choose the default resolution of a conflict where multiple ops concurrently assigned values to the same key). For opIDs with the same counter component, we currently rely on lexicographic ordering of actor IDs. Switching this to lexicographic order of hashes would work. Better might be to ensure the order of changes in the document's change list is consistent across all peers; then opID comparison could be defined in terms of comparing indices into that list. Consistent change order would probably have to be a deterministic topological sort of the change dependency graph.

For querying values at a given set of heads, one approach might be to first translate the heads into a set of ranges of indexes in the table of changes. Then a visibility check for an op would mean checking whether the op's change index falls within one of those ranges.

## Cost of all those hashes

I worry about the cost of storing and/or recomputing all the hashes, and keeping the hashes in memory. For the paper editing trace with 260k changes, the hashes alone add up to more than 8MB; for comparison, the entire Automerge-unstable memory consumption for the same file is [23MB](https://automerge.org/blog/automerge-2/), so the hashes could easily end up being a large fraction of total memory use. Moreover, recomputing the hash graph on document load might be prohibitively slow.

At the moment, the use of actor IDs allows us to skip the hash graph computation in many cases: we can just store the hashes for the heads as part of the document, and as long as we don't need to refer back to any historical hashes further back in the history, we can load the document, make and receive changes without ever computing the hash graph. However, if most operations refer to opIDs with hashes that might be anywhere in the prior history, we refer to old hashes much more often, and so we would have to compute the hash graph much sooner.

If we only store, say, 1% of hashes, then the cost of storing them would be acceptable, and if we store those hashes as part of the document, we wouldn't need to recompute the hash graph on changes that refer only to stored hashes. I suspect this kind of sparse storage could be achieved using a similar technique to point 3 in the section on encoding individual changes. For example, if we have a linear sequence of changes `a <-- b <-- c <-- d <-- e <-- ...` (where each change depends on only one prior change), and `e` is in the 1% of changes we store, then we could unambiguously refer to `b` as "3 changes before `e`" instead of using the hash of `b`. In addition to 1% of changes in linear sequences, we would also need to explicitly store the heads of all branches, and any hashes of changes that are a dependency of a merge. (Alternatively, we could do a topological sort of the dependency graph, but that would be more complex).

All peers would need to use the same deterministic rule to decide which of the hashes to store, but this could simply be "hashes of all changes with a sequence number divisible by 100" or similar. I suspect that would make the cost of the hashes low enough that it's not a problem.

[thanks to baby for sleeping for long enough that I was able to type this up!]

**Comment by burdiyan** [comment]
Are there any plans to pursue the removal of actor IDs any time soon?

**Comment by burdiyan** [comment]
> I worry about the cost of storing and/or recomputing all the hashes, and keeping the hashes in memory. For the paper editing trace with 260k changes, the hashes alone add up to more than 8MB; for comparison, the entire Automerge-unstable memory consumption for the same file is [23MB](https://automerge.org/blog/automerge-2/), so the hashes could easily end up being a large fraction of total memory use. Moreover, recomputing the hash graph on document load might be prohibitively slow.
> 
> At the moment, the use of actor IDs allows us to skip the hash graph computation in many cases: we can just store the hashes for the heads as part of the document, and as long as we don't need to refer back to any historical hashes further back in the history, we can load the document, make and receive changes without ever computing the hash graph. However, if most operations refer to opIDs with hashes that might be anywhere in the prior history, we refer to old hashes much more often, and so we would have to compute the hash graph much sooner.

One idea that I had (which may well be a total nonsense) for storing lots of hashes more compactly is using prefixes of a hash mapped to a list of data that shares this prefix, similar to how collisions are handled in a hash table. E.g. we could take the first 64 bits of the hash, cast it to a number (so it doesn't consume heap memory) and use those numbers as map keys.

64 bits should probably be enough for collisions to happen very rarely, but it would be worth to test it on a real benchmark data.

It still would mean that in order to truly say if you have the data of the hash you'd have to rehash the data again to see if it matches for real, or only its prefix matched, so maybe keeping some bounded LRU cache of full hashes could help mitigate the problem. Care probably must be taken so it doesn't become an attack vector when someone could make your device rehash all the data by probing all the different hashes on you.

**Comment by BrianHicks** [comment]
Regarding this:

> All peers would need to use the same deterministic rule to decide which of the hashes to store, but this could simply be "hashes of all changes with a sequence number divisible by 100" or similar. I suspect that would make the cost of the hashes low enough that it's not a problem.

What stops a scheme like this from only storing hashes for the root and any multi-parent node, and treating indexes as the count since the last one of those operations? Would that make some other desirable operations very difficult?

**Reply by ept** [reply]
The problem is that any change anywhere in the history could be used as branching-off point for a new branch. When you receive the first change on a new branch, its parent could be any existing hash, and the challenge is finding that hash so that you know where the new branch needs to go. It would be nice to be able to do this without having to recompute all the hashes in the entire history. With actorIds and sequence numbers, one option would be to record the parent's actorId+seqNum in the first change of the new branch, to use them to locate the branching-off point, and then to use the hash only as a consistency check. But if we don't have actorIds any more, this becomes more challenging.

**Reply by ept** [reply]
Count since the last merge is also ambiguous when you have branching, since you don't know which branch to follow. It would be possible to use count of changes _prior_ to a merge (multi-parent change), but counts into the future are unreliable.

**Comment by zaqqwerty** [comment]
Hi all, thanks for the discussion - what would be the recommended way to associate users to their change history in a document?

Use case: we are beginning to use Automerge repos to keep document edits in our webapp synced between users, and want to maintain history of which authenticated user made which changes to the docs.  Seems like the way to do this would be to maintain a table binding actor IDs to user ID, but if removing actor IDs is being suggested, I'm assuming there is a better recommended approach.  Thanks!

**Comment by svdvonde** [comment]
Hi, I was pointed here via the Automerge discord. I'm a user of Automerge for research proof-of-concepts, so my use of actor ids is kind of esoteric. In essence, I have to inconveniently manipulate actor ids to get identical document outcomes when applying patches to documents.

I've been using actor ids in two ways.

1. I have to 'recreate'/'mirror' changes to one document in a different document. I use actor ids to force a merge conflict to be resolved one way, such that the 'winning' value in one document is also winning in the mirrored document.
3. I have a homebrew synchronisation layer where two peers independently apply a set of operations to their local document, and they are expected to have _exactly_ the same outcome (as in: exact same document with all internal hashes being identical). Since the internal hashes are computed based on (among other things) actor ids, if I want the resulting documents to be hash-identical, I actually communicate actor ids between peers such that one peer can 'impersonate' the other peer.
